overlord/configstate/configcore/tmpfs.go

```diff
@@ -0,0 +1,127 @@
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/sysconfig"
+)
+
+const (
+	mntStaticOptions         = "mode=1777,strictatime,nosuid,nodev"
+	tmpfsService             = "tmp.mount"
+	tmpfsMountPoint          = "/tmp"
+	tmpMntServOverrideSubDir = "tmp.mount.d"
+	tmpMntServOverrideFile   = "override.conf"
+)
+
+func init() {
+	// add supported configuration of this module
+	supportedConfigurations["core.tmpfs.size"] = true
+}
+
+// Regex matches what is specified by tmpfs(5) for the size option
+var validTmpfsSize = regexp.MustCompile(`^[0-9]+[kmgKMG%]?$`).MatchString
+
+func validateTmpfsSettings(tr config.ConfGetter) error {
+	tmpfsSz, err := coreCfg(tr, "tmpfs.size")
+	if err != nil {
+		return err
+	}
+	if tmpfsSz == "" {
+		return nil
+	}
+	if !validTmpfsSize(tmpfsSz) {
+		return fmt.Errorf("cannot set tmpfs size %q: invalid size", tmpfsSz)
+	}
+
+	return nil
+}
+
+func handleTmpfsConfiguration(_ sysconfig.Device, tr config.ConfGetter,
+	opts *fsOnlyContext) error {
+
+	tmpfsSz, err := coreCfg(tr, "tmpfs.size")
+	if err != nil {
+		return err
+	}
+
+	// Create override configuration file for tmp.mount service
+
+	// Create /etc/systemd/system/tmp.mount.d if needed
+	var overrDir string
+	if opts == nil {
+		// runtime system
+		overrDir = dirs.SnapServicesDir
+	} else {
+		overrDir = dirs.SnapServicesDirUnder(opts.RootDir)
+	}
+	overrDir = filepath.Join(overrDir, tmpMntServOverrideSubDir)
+
+	// Write service config override if needed
+	options := mntStaticOptions
+	dirContent := make(map[string]osutil.FileState, 1)
+	if tmpfsSz != "" {
+		if err := os.MkdirAll(overrDir, 0755); err != nil {
+			return err
+		}
+		options = fmt.Sprintf("%s,size=%s", options, tmpfsSz)
+		content := fmt.Sprintf("[Mount]\nOptions=%s\n", options)
+		dirContent[tmpMntServOverrideFile] = &osutil.MemoryFileState{
+			Content: []byte(content),
+			Mode:    0644,
+		}
+	} else {
+		// Use default tmpfs size if empty setting (50%, see tmpfs(5))
+		options = fmt.Sprintf("%s,size=50%%", options)
+	}
+	glob := tmpMntServOverrideFile
+	changed, removed, err := osutil.EnsureDirState(overrDir, glob, dirContent)
```

Comment:
I wonder if we should have a defer here that makes a copy of the existing file and reverts on err?

Reply:
Maybe, but I see a couple of problems with that:
1. We do not know what EnsureDirState() has done before failing
2. EnsureDirState() just writes/removes files/dirs, so if it fails probably our revert will fail as well as it will perform also file accesses

So I think that if some revert needs to happen probably it should be inside EnsureDirState().